analytics_data_api/v0/models.py

```diff
@@ -220,3 +223,64 @@ class Meta:
     def get_course_user(cls, course_id, username):
         return cls.search().query('term', course_id=course_id).query(
             'term', username=username).execute()
+
+
+class ModuleEngagement(models.Model):
+    """User interactions with entities within the courseware."""
+
+    course_id = models.CharField(db_index=True, max_length=255)
+    username = models.CharField(max_length=255)
+    date = models.DateTimeField()
+    # This will be one of "problem", "video" or "forum"
+    entity_type = models.CharField(max_length=255)
+    # For problems this will be the usage key, for videos it will be the html encoded module ID,
+    # for forums it will be the commentable_id
+    entity_id = models.CharField(max_length=255)
+    # A description of what interaction occurred.
+    event = models.CharField(max_length=255)
+    # The number of times the user interacted with this entity in this way on this day.
+    count = models.IntegerField()
+
+    class Meta(object):
+        db_table = 'module_engagement'
+
+
+class ModuleEngagementTimelineManager(models.Manager):
+    """
+    Modifies the ModuleEngagement queryset to aggregate engagement data for
+    the learner engagement timeline.
+    """
+    def get_timelines(self, course_id, username):
+        queryset = ModuleEngagement.objects.all().filter(course_id=course_id, username=username)\
+            .values('date', 'entity_type', 'event')\
+            .annotate(count=Sum('count'))\
+            .order_by('date')
+
+        timelines = []
+
+        for key, group in groupby(queryset, lambda x: (x['date'])):
+            # Iterate over groups and create a single item with engagement data
+            item = {
+                u'date': key,
+            }
+            for engagement in group:
+                entity_type = engagement_entity_types.SINGULAR_TO_PLURAL[engagement['entity_type']]
+                engagement_type = '{}_{}'.format(entity_type, engagement['event'])
+                count = item.get(engagement_type, 0)
+                count += engagement['count']
+                item[engagement_type] = count
+            timelines.append(item)
+
+        return timelines
+
+
+class ModuleEngagementTimeline(models.Model):
+    """
+    Learner engagement timeline data.
+    """
+    date = models.DateField()
+    problems_attempted = models.IntegerField()
+    problems_completed = models.IntegerField()
+    discussions_contributed = models.IntegerField()
+    videos_viewed = models.IntegerField()
+    objects = ModuleEngagementTimelineManager()
```

Comment:
I'm used to the django convention of the manager belonging to the model it manages. If the `ModulelEngagementTimelineManager` were to belong to the `ModuleEngagement` class, would we be able to get rid of this model?

I'm also skeptical of having a model that doesn't map to a database table.